rfcs/0001-extended-lookup-tables.md

```diff
@@ -0,0 +1,503 @@
+
+We propose extending each table with an extra column, which we will call the
+`TableID` column, which associates a different number with each table. Since
+the sizes of all tables must be known at constraint system time, we can treat
+this as a fixed column. For example, this converts a lookup `(a, b, c)` in the
+table with ID `1` into a lookup of `(a, b, c, 1)` in the combined table, and
+ensures that any value `(x, y, z)` from table ID `2` will not match that
+lookup, since its representative is (x, y, z, 2)`.
+
+To avoid any interaction between the `TableID` column and any other data in
+the tables, we use `TableID * joint_combiner^max_joint_size` to mix in the
+`TableID`, where `max_joint_size` is the maximum length of any lookup or table
+value. This also simplifies the 'joining' of the `table_id` polynomial
+commitment, since we can straightforwardly scale it by the constant
+`joint_combiner^max_joint_size` to find its contribution to every row in the
+combined table.
+
+### Sampling the `joint_combiner`
+
+Currently, the `joint_combiner` is sampled using randomness that depends on the
+witness. However, where we have runtime or side-loaded tables, a malicious
+prover may be able to select values in those tables that absue knowledge of
+`joint_combiner` to create collisions.
+
+For example, to maliciously insert a value `(x, y, z)` for table with ID 1, the
+prover could select the value `x + y * joint_combiner + z * joint_combiner^2 +
+-1 * joint_combiner^max_joint_size` for an entry in table 2.
+
+Thus, we need to ensure that the `joint_combiner` depends on the values in
+runtime or side-loaded tables, so that the values in these tables cannot be
+chosen using knowledge of the `joint_combiner`. To do this, we must use the
+commitments to the values in these tables as part of the source of randomness
+for `joint_combiner` in the prover and verifier.
+
+In particular, this means that we must have a commitment per column for each
+type of table, so that the verifier can confirm the correct construction of the
+combined table.
+
+### Representing the combined fixed table
+
+The fixed table can be constructed at constraint system generation time, by
+extending all of the constituent tables to the same width, generating the
+appropriate table IDs array, and concatenating all of the tables. Concretely:
+
+Let `t[i][j][k]` be the `k`th element of the `j`th entry in the `i`th fixed
+table.
+Let `W` be the maximum width of all of the `t[i][j]`s.
+For any `t[i][j]` whose width `w` is less than `W`, pad it to width `W` by
+setting `t[i][j][k] = 0` for all `w < k <= W`.
+Form the combined table by concatenating all of these tables
+`FixedTable = t[0] || t[1] || t[2] || ...`
+and store in `TableID` the table ID that the corresponding row came from.
+Then, for every `x` and `y`, we have
+```
+FixedTable[x][k] = t[i][j][k]`
+TableID[x] = i
+where x = len(t[0]) + len(t[1]) + ... + len(t[i-1]) + j
```

Comment:
This is formalising the concatenation `t[0] || t[1] || ...`, so that `output_row` represents the current row in the concatenated table. `len(t[id])` is supposed to be the number of rows in the table `t[id]`, so we might have e.g.
```
t[0] = [[0, 0, 0], [0, 1, 1], [1, 0, 1], [1, 1, 1]] // len(t[0]) = 4
t[1] = [[0], [1], [2], [3], [4], [5], [6], [7]] // len(t[1]) = 8
let output_row = 7
FixedTable[output_row][col] = t[1][3][col]
output_row = len(t[0]) + 3 = 4 + 3 = 7
```

Any suggestions on how to make this clearer?